pkg/apis/pipeline/v1alpha1/build_gcs_resource.go

```diff
@@ -0,0 +1,179 @@
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "Location"):
+			location = param.Value
+		case strings.EqualFold(param.Name, "DestinationDir"):
+			destDir = param.Value
+		case strings.EqualFold(param.Name, "ArtifactType"):
+			var err error
+			aType, err = getArtifactType(param.Value)
+			if err != nil {
+				return nil, fmt.Errorf("BuildGCSResource %s : %s", r.Name, err)
+			}
+		}
+	}
+	if location == "" {
+		return nil, fmt.Errorf("BuildGCSResource: Need Location to be specified in order to create BuildGCS resource %s", r.Name)
+	}
+	if aType == EmptyArtifactType {
+		return nil, fmt.Errorf("BuildGCSResource: Need ArtifactType to be specified in order to fetch BuildGCS resource %s", r.Name)
+	}
+	return &BuildGCSResource{
+		Name:           r.Name,
+		Type:           r.Spec.Type,
+		Location:       location,
+		DestinationDir: destDir,
+		ArtifactType:   aType,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s BuildGCSResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s BuildGCSResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeStorage
+}
+
+// GetParams get params
+func (s *BuildGCSResource) GetParams() []Param { return []Param{} }
+
+// GetSecretParams returns the resource secret params
+func (s *BuildGCSResource) GetSecretParams() []SecretParam { return nil }
+
+// Replacements is used for template replacement on an GCSResource inside of a Taskrun.
+func (s *BuildGCSResource) Replacements() map[string]string {
+	return map[string]string{
+		"name":     s.Name,
+		"type":     string(s.Type),
+		"location": s.Location,
+	}
+}
+
+// SetDestinationDirectory sets the destination directory at runtime like where is the resource going to be copied to
+func (s *BuildGCSResource) SetDestinationDirectory(destDir string) { s.DestinationDir = destDir }
+
+// GetDownloadContainerSpec returns an array of container specs to download gcs storage object
+func (s *BuildGCSResource) GetDownloadContainerSpec() ([]corev1.Container, error) {
+	if s.DestinationDir == "" {
+		return nil, fmt.Errorf("BuildGCSResource: Expect Destination Directory param to be set %s", s.Name)
+	}
+	args := []string{"--type", string(s.ArtifactType), "--location", s.Location}
+	// dest_dir is the destination directory for GCS files to be copies"
+	if s.DestinationDir != "" {
+		args = append(args, "--dest_dir", s.DestinationDir)
+	}
+
+	return []corev1.Container{
+		CreateDirContainer(s.Name, s.DestinationDir), {
+			Name:  fmt.Sprintf("storage-fetch-%s", s.Name),
+			Image: *buildGCSFetcherImage,
+			Args:  args,
+		}}, nil
+}
+
+// GetUploadContainerSpec gets container spec for gcs resource to be uploaded like
+// set environment variable from secret params and set volume mounts for those secrets
+func (s *BuildGCSResource) GetUploadContainerSpec() ([]corev1.Container, error) {
+	if s.ArtifactType != GCSManifest {
```

Comment:
Does it mean that upload container spec only supports Manifest type ? If so why?

Reply:
yes for now the looking at the [code](https://github.com/GoogleCloudPlatform/cloud-builders/blob/master/gcs-fetcher/cmd/gcs-uploader/main.go#L38) for gcs-uploader, the parameter mentions,` --location` should point to manifest file.
There are no check to ensure that in the code, however I followed along the code path and if the uploader does expect a manifest object.
[here](https://github.com/GoogleCloudPlatform/cloud-builders/blob/master/gcs-fetcher/pkg/uploader/uploader.go#L61)
```
type Uploader struct {
	gcs                    GCS
	os                     OS
	bucket, manifestObject string

	manifest                 sync.Map
	totalBytes, bytesSkipped int64
}

func New(ctx context.Context, gcs GCS, os OS, bucket, manifestObject string, numWorkers int) *Uploader {
...
}
```

Reply:
Cool. 
If possible can you add the above description as comment in code?